### PR TITLE
More advanced and thorough examples in docs

### PR DESCRIPTION
**Pull request by bmcfee**
This PR implements #41 #40 -- adding examples for handling NPZ files and parallelism.

### CODE REVIEW

**Comment by bmcfee** [comment]
Probably a good idea to include a fully worked example in the docs, eh?  Eg data augmentation on iris via sklearn.

**Comment by bmcfee** [comment]
Tagging @justinsalamon

File: docs/example2.rst [hunk 1]
```diff
@@ -48,28 +48,49 @@ As a concrete example, we can simulate a mixture of noisy streams with differing
...
 
         # Instantiate a linear classifier
         estimator = SGDClassifier()
 
         # Build a collection of Streamers with different noise scales
         streams = [Streamer(noisy_samples, X[train], Y[train], sigma=sigma)
-                   for sigma in [0.5, 1.0, 2.0, 4.0]]
+                   for sigma in [0, 0.5, 1.0, 2.0, 4.0]]
 
         # Build a mux stream, keeping only 2 streams alive at once
```
Comment (cjacoby) [inline docs/example2.rst]: Here it says 2 streams, below 3. I assume this is not what you mean?

File: docs/example2.rst [hunk 2]
```diff
@@ -48,28 +48,49 @@ As a concrete example, we can simulate a mixture of noisy streams with differing
...
         mux_stream = Mux(streams,
-                         k=2,    # Keep 2 streams alive at once
-                         lam=16) # Use a poisson rate of 16
+                         k=3,    # Keep 3 streams alive at once
+                         lam=64) # Use a poisson rate of 64
```
Comment (cjacoby) [inline docs/example2.rst]: In the below text, you say "...`lam=16`, but it's now 64. Plzfix.

Comment (bmcfee) [inline docs/example2.rst]: Done.

**Comment by justinsalamon** [comment]
@bmcfee happy to pitch in once im back, if needed.

File: docs/example3.rst [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        # so that memory is released when the streamer is deactivated
+        with np.load(npz_file) as data:
+            # How many rows are in the data?
+            # We assume that data['Y'] has the same length
+            n_total = len(data['X'])
+
+            while True:
+                # Compute the index offset
+                idx = np.random.randint(n_total - n)
+                yield dict(X=data['X'][idx:idx+n],
```
Comment (ejhumphrey) [inline docs/example3.rst]: IIRC, this kind of slicing will (should?) cause memory leaks because the dictionary inherits the identifier of `data`, despite using the context manager. I believe [this demonstrates the behavior.](https://github.com/ejhumphrey/pescador-wtf/blob/master/test_memleak.py#L104)

I think the best practice here is to force the data copy before handing an ndarray to the dictionary, like `dict(X=np.array(data['X'][idx:idx + n]...`.

note: it makes me *wicked* uncomfortable how easy this is to overlook, but I don't have any answers I like. There are three that jump out at me:

- `Streamer` objects have a keyword argument like `asarray=True` which, by default, forces values in dictionaries to be reallocated as new arrays. I _think_ this would do the right thing, but I'm maybe 65% sure of that. Major win, the default behavior is sane; potential drawback, how detrimental to performance is it to blindly reallocate all matrix data?
- We provide an `@asarray` decorator, that one can apply to generators to enforce the data it yields is properly reallocated, allowing Python to fully clean up after a stream when it is deactivated. Major disadvantage here is that this should be the default behavior, but decorators are explicit add-ons, and like to be forgotten.
- We provide a `Sample` object, as a subclass of `dict` that does the right thing with the data it is handed. Again, not exactly default behavior, and is starting to feel slightly over-engineered.

any other ideas?

nit: there should be whitespace around the `+`s

**Comment by bmcfee** [comment]
> IIRC, this kind of slicing will (should?) cause memory leaks because the dictionary inherits the identifier of data, despite using the context manager. I believe this demonstrates the behavior.

I'm not sure that's a leak exactly - the reference is still active, so the GC won't (and shouldn't) free it.

However, once the context manager exits AND the last reference to the array is lost, then the GC should collect it.  If it's not, that's a bug in python.

It's worth mentioning that a buffered streamer does force the copy, so that could be an extra sanity check.

(In case it's not obvious, I really don't like doing explicit copies just to avoid apparent GC weirdness.)

**Comment by bmcfee** [comment]
> I have also thought about making a sample object (I've done this in some of my own work), so that additionally the Sample object could use __slots__ = ['X', 'Y'], and thereby save a lot of memory.

The downside of this is that you'd be stuck with only the two pre-determined variable names. I sometimes have streamers that generate many more variables (eg for multitask learning for different Y's, or cross-modal learning for different X's).

Would it be possible to do something that works like a named tuple (dynamically set, but fixed fields), that you could still use in the way you describe above?

File: docs/example3.rst (hunk 3 above)
Comment (bmcfee) [inline docs/example3.rst]: I've thought a bit more about this, and I think it's best that we not over-complicate things.  It's already complicated enough to use, and I don't like the idea of doing reallocation behind the scenes.

FWIW, this thread is helpful: https://mail.scipy.org/pipermail/numpy-discussion/2015-December/074376.html

TLDR: `npy` files support [mmap mode](https://docs.scipy.org/doc/numpy/reference/generated/numpy.load.html) while `npz` files do not.  If you care about memory leaking here, don't use `npz` files.  If you want memory mapping with structured data, use `pytables` instead.

I think the sane thing to do here is to drop this example, and note the memory mapping problem in the docs.

**Comment by bmcfee** [comment]
I updated the npz/npy example again to discuss memory mapping issues, and while I was at it, fixed the build on RTD.  @ejhumphrey care to take another pass on this?

**Comment by ejhumphrey** [comment]
looks great 🎉
